File: scripts/parse_crane_csv.py
"""
parse_crane_csv.py

Reference implementation of the LTM 1130-5.1 duty-chart CSV parser.
Converts the hand-transcribed CSV (one block per counterweight/jib
configuration, 5 spec rows + blank + header row(s) + data rows,
repeated) into the CraneModel JSON shape defined in lib/types.ts.

FILE FORMAT (confirmed against the actual uploaded CSV):
  - Encoding is Mac OS Roman (not UTF-8/Latin-1) — degree symbols in
    jib-angle header rows decode incorrectly under any other codec.
  - Each config block:
      row 0: "Range", "<min>_<max>"
      row 1: "Outriggers Fully Deployed", "Yes"/"No"
      row 2: "Slew Radius", "<degrees>"
      row 3: "Counterweight", "<tonnes>"
      row 4: "FlyJib", "No" OR "<jib length in m>"
      row 5: blank
      row 6: boom-length header row (one column per boom, or per
             boom+angle triplet if a jib is fitted)
      row 7 (jib configs only): jib-angle header row (0/20/40 deg)
      data rows: col0 = radius, remaining cols = capacity at that
             (boom[, angle]) column, blank = not achievable
      terminated by a blank row or EOF
  - Boom-length header cells may carry a unit suffix ("12.7 m" vs
    "12.7") inconsistently between blocks — parser strips this.
  - A boom-length header cell suffixed "*" denotes a separate
    "over rear" column for that same boom length (vs the default
    "over front") — this ONLY occurred on the 12.7m boom in the
    42t/no-jib block in the file received; parser handles it
    generically wherever it appears.
  - Jib blocks repeat each boom length 3× (one column per 0/20/40
    degree offset); the jib-angle row underneath the boom-length row
    supplies the offset for each column.

MERGING: a jib block's capacities are merged into the matching
BoomConfig (same counterweight + boomLengthM) from that counterweight's
plain (no-jib) block, as an entry in that BoomConfig's `jibs[]` array.
"""
import json
import re

SRC = "/mnt/user-data/uploads/LTM1130-5_1.csv"
OUT = "/home/claude/lift-planner/data/cranes/ltm-1130-5.1.json"

NUM_RE = re.compile(r'^(-?\d+(?:\.\d+)?)')


def load_rows(path):
    with open(path, "rb") as f:
        raw = f.read()
    text = raw.decode("mac_roman")
    lines = text.split("\r\n")
    return [line.split(",") for line in lines]


def cell(rows, r, c):
    if r < 0 or r >= len(rows):
        return ""
    row = rows[r]
    if c < 0 or c >= len(row):
        return ""
    return row[c].strip()


def is_blank_row(rows, r):
    row = rows[r] if r < len(rows) else []
    return all(x.strip() == "" for x in row)


def parse_boom_header_cell(raw_val):
    v = raw_val.strip()
    if v == "":
        return None
    orientation = "front"
    if v.endswith("*"):
        orientation = "rear"
        v = v[:-1].strip()
    m = NUM_RE.match(v)
    if not m:
        return None
    return float(m.group(1)), orientation


def parse_angle_header_cell(raw_val):
    v = raw_val.strip().rstrip("\xa1\u00b0 ")
    if v == "":
        return None
    m = NUM_RE.match(v)
    if not m:
        return None
    return int(float(m.group(1)))


def parse_blocks(rows):
    i = 0
    n = len(rows)
    blocks = []
    while i < n:
        if cell(rows, i, 0) == "Range":
            outriggers_raw = cell(rows, i + 1, 1)
            slew_raw = cell(rows, i + 2, 1)
            cw = float(cell(rows, i + 3, 1))
            flyjib_raw = cell(rows, i + 4, 1)
            has_jib = flyjib_raw not in ("No", "")
            jib_len = float(flyjib_raw) if has_jib else None
            header_start = i + 6
            if has_jib:
                boom_header = rows[header_start]
                angle_header = rows[header_start + 1]
                data_start = header_start + 2
            else:
                boom_header = rows[header_start]
                angle_header = None
                data_start = header_start + 1

            columns = []
            for c in range(1, len(boom_header)):
                parsed = parse_boom_header_cell(boom_header[c] if c < len(boom_header) else "")
                if parsed is None:
                    continue
                boom_len, orientation = parsed
                offset = None
                if angle_header is not None:
                    offset = parse_angle_header_cell(angle_header[c] if c < len(angle_header) else "")
                    if offset is None:
                        continue
                columns.append({"col": c, "boomLengthM": boom_len, "orientation": orientation, "offsetDeg": offset})

            r = data_start
            data_rows = []
            while r < n and not is_blank_row(rows, r) and cell(rows, r, 0) != "Range":
                radius_val = cell(rows, r, 0)
                if radius_val != "":
                    try:
                        radius = float(radius_val)
                        row_vals = {}
                        for colinfo in columns:
                            v = cell(rows, r, colinfo["col"])
                            if v != "":
                                try:
                                    row_vals[colinfo["col"]] = float(v)
                                except ValueError:
                                    pass
                        data_rows.append((radius, row_vals))
                    except ValueError:
                        pass
                r += 1

            blocks.append({
                "outriggersFullyDeployed": outriggers_raw == "Yes",
                "slewRadiusDeg": float(slew_raw) if slew_raw else None,
                "counterweightTonnes": cw,
                "jibLengthM": jib_len,
                "columns": columns,
                "dataRows": data_rows,
            })
            i = r + 1
        else:
            i += 1
    return blocks


def build_crane_model(blocks):
    plain_blocks = [b for b in blocks if b["jibLengthM"] is None]
    jib_blocks = [b for b in blocks if b["jibLengthM"] is not None]

    counterweights = []
    for b in plain_blocks:
        boom_configs = []
        # group columns by (boomLengthM, orientation)
        for colinfo in b["columns"]:
            capacities = []
            for radius, vals in b["dataRows"]:
                v = vals.get(colinfo["col"])
                if v is not None:
                    capacities.append({"radiusM": radius, "capacityTonnes": v})
            boom_entry = {
                "boomLengthM": colinfo["boomLengthM"],
                "capacities": capacities,
            }
            if colinfo["orientation"] == "rear":
                boom_entry["orientation"] = "rear"
            boom_configs.append(boom_entry)

        # attach jibs for this counterweight from matching jib blocks
        matching_jib_blocks = [jb for jb in jib_blocks if jb["counterweightTonnes"] == b["counterweightTonnes"]]
        for jb in matching_jib_blocks:
            # group jib columns by boomLengthM -> list of offsetDeg columns
            booms_in_jib = {}
            for colinfo in jb["columns"]:
                booms_in_jib.setdefault(colinfo["boomLengthM"], []).append(colinfo)

            for boom_len, col_list in booms_in_jib.items():
                jib_capacities_by_offset = {}
                for colinfo in col_list:
                    capacities = []
                    for radius, vals in jb["dataRows"]:
                        v = vals.get(colinfo["col"])
                        if v is not None:
                            capacities.append({"radiusM": radius, "capacityTonnes": v})
                    jib_capacities_by_offset[colinfo["offsetDeg"]] = capacities

                jib_config_entries = [
                    {"jibLengthM": jb["jibLengthM"], "offsetDeg": offset, "capacities": caps}
                    for offset, caps in sorted(jib_capacities_by_offset.items())
                ]

                # find matching boom_config (front orientation, since jib
                # columns in this file never carry an asterisk)
                target = next((bc for bc in boom_configs if bc["boomLengthM"] == boom_len and bc.get("orientation") is None), None)
                if target is None:
                    # boom length only appears in the jib chart, not the
                    # plain chart — create a standalone entry for it
                    target = {"boomLengthM": boom_len, "capacities": []}
                    boom_configs.append(target)
                target.setdefault("jibs", []).extend(jib_config_entries)

        cw_label = f"{b['counterweightTonnes']:g}".replace('.', '_')
        counterweights.append({
            "id": f"cw_{cw_label}t",
            "weightTonnes": b["counterweightTonnes"],
            "outriggersFullyDeployed": b["outriggersFullyDeployed"],
            "slewRadiusDeg": b["slewRadiusDeg"],
            "boomConfigs": boom_configs,
        })

    return {
        "craneModel": "Liebherr LTM 1130-5.1",
        "riggedWeightTonnes": 60,
        "baseCounterweightTonnes": 9,
        "maxLinePullTonnes": 9.34,
        "totalRopeLengthM": 250,
        "hookBlocks": [
            {"id": "hb_110t", "label": "110t / 7-sheave", "ratedCapacityTonnes": 110, "sheaves": 7, "maxLines": 14, "weightKg": 1240},
            {"id": "hb_90_2t", "label": "90.2t / 5-sheave", "ratedCapacityTonnes": 90.2, "sheaves": 5, "maxLines": 11, "weightKg": 900},
            {"id": "hb_59_1t", "label": "59.1t / 3-sheave", "ratedCapacityTonnes": 59.1, "sheaves": 3, "maxLines": 7, "weightKg": 700},
            {"id": "hb_26_1t", "label": "26.1t / 1-sheave", "ratedCapacityTonnes": 26.1, "sheaves": 1, "maxLines": 3, "weightKg": 450},
            {"id": "hb_8_8t", "label": "8.8t / single line", "ratedCapacityTonnes": 8.8, "sheaves": 0, "maxLines": 1, "weightKg": 250},
        ],
        "counterweights": counterweights,
    }


if __name__ == "__main__":
    rows = load_rows(SRC)
    blocks = parse_blocks(rows)
    print(f"Parsed {len(blocks)} config blocks from source CSV")
    model = build_crane_model(blocks)
    print(f"Built CraneModel with {len(model['counterweights'])} counterweight configs")
    for cw in model["counterweights"]:
        n_jibs = sum(len(bc.get("jibs", [])) for bc in cw["boomConfigs"])
        print(f"  {cw['weightTonnes']}t: {len(cw['boomConfigs'])} boom configs, {n_jibs} jib configs attached")
    with open(OUT, "w") as f:
        json.dump(model, f, indent=2)
    print(f"Wrote {OUT}")

File: lib/data-prep/parseCraneCsv.ts
// ============================================================
// lib/data-prep/parseCraneCsv.ts
//
// Converts a hand-transcribed crane duty-chart CSV into the
// CraneModel shape defined in lib/types.ts. This is a build-time
// / data-prep utility (run via a script, not shipped to the
// browser) — its output is what actually gets bundled as
// data/cranes/*.json.
//
// This is a direct TypeScript port of scripts/parse_crane_csv.py,
// which was used to validate the parsing logic against the real
// LTM 1130-5.1 CSV (spot-checked against two independently-known
// real values before being trusted). Keep the two in sync if you
// change the algorithm, or retire the Python version once this is
// wired into the actual Next.js build pipeline.
//
// FILE FORMAT (confirmed against a real hand-transcribed CSV):
//   - Encoding is commonly Mac OS Roman when exported from Excel on
//     macOS — degree symbols in jib-angle header rows will decode
//     incorrectly under UTF-8/Latin-1. Detect/handle both.
//   - Each config block, top to bottom:
//       row 0: "Range", "<min>_<max>"
//       row 1: "Outriggers Fully Deployed", "Yes"/"No"
//       row 2: "Slew Radius", "<degrees>"
//       row 3: "Counterweight", "<tonnes>"
//       row 4: "FlyJib", "No" OR "<jib length in m>"
//       row 5: blank
//       row 6: boom-length header row (one column per boom, or per
//              boom+angle triplet if a jib is fitted for this block)
//       row 7 (jib configs only): jib-angle header row (0/20/40 deg)
//       data rows: col0 = radius, remaining cols = capacity at that
//              (boom[, angle]) column; blank = not achievable
//       block ends at a blank row, "Range", or EOF
//   - Boom-length header cells may carry an inconsistent unit suffix
//     ("12.7 m" vs "12.7") between blocks — always strip non-numeric
//     trailing characters rather than assuming a fixed format.
//   - A boom-length header cell suffixed "*" denotes a distinct
//     "over rear" column for that same boom length (default is
//     "over front" when unmarked). Confirmed real case: LTM 1130-5.1,
//     42t counterweight, 12.7m boom, no jib.
//   - Jib blocks repeat each boom length once per published offset
//     angle (typically 0/20/40°); the jib-angle header row supplies
//     the offset for each column.
//
// MERGING: capacities from a jib block are merged into the matching
// BoomConfig (same counterweight + boomLengthM, 'front' orientation)
// taken from that counterweight's plain (no-jib) block, added as an
// entry in that BoomConfig's `jibs[]` array. If a jib block references
// a boom length that has no corresponding plain-chart entry, a
// standalone BoomConfig (with empty `capacities`) is created to hold
// the jib entry rather than silently dropping it.
// ============================================================

import type {
  CraneModel,
  CounterweightConfig,
  BoomConfig,
  JibConfig,
  RadiusCapacityPoint,
  HookBlockSpec,
} from '../types';

interface ParsedColumn {
  col: number;
  boomLengthM: number;
  orientation: 'front' | 'rear';
  offsetDeg: number | null; // null for non-jib blocks
}

interface ParsedBlock {
  outriggersFullyDeployed: boolean;
  slewRadiusDeg: number | null;
  counterweightTonnes: number;
  jibLengthM: number | null;
  columns: ParsedColumn[];
  dataRows: Array<{ radiusM: number; values: Map<number, number> }>;
}

const NUM_PREFIX = /^(-?\d+(?:\.\d+)?)/;

function cellAt(rows: string[][], r: number, c: number): string {
  if (r < 0 || r >= rows.length) return '';
  const row = rows[r];
  if (c < 0 || c >= row.length) return '';
  return row[c].trim();
}

function isBlankRow(rows: string[][], r: number): boolean {
  const row = rows[r] ?? [];
  return row.every((x) => x.trim() === '');
}

function parseBoomHeaderCell(raw: string): { boomLengthM: number; orientation: 'front' | 'rear' } | null {
  let v = raw.trim();
  if (v === '') return null;
  let orientation: 'front' | 'rear' = 'front';
  if (v.endsWith('*')) {
    orientation = 'rear';
    v = v.slice(0, -1).trim();
  }
  const m = NUM_PREFIX.exec(v);
  if (!m) return null;
  return { boomLengthM: parseFloat(m[1]), orientation };
}

function parseAngleHeaderCell(raw: string): number | null {
  // strip degree-symbol variants across encodings (° U+00B0, or the
  // mis-decoded \xA1 seen under Mac OS Roman misread as Latin-1/UTF-8)
  const v = raw.trim().replace(/[\u00b0\u00a1\s]+$/g, '');
  if (v === '') return null;
  const m = NUM_PREFIX.exec(v);
  if (!m) return null;
  return Math.round(parseFloat(m[1]));
}

/**
 * Splits raw CSV text into a grid of string cells. Assumes no quoted
 * commas (true of every duty-chart CSV seen so far — verify this
 * holds before reusing on a new source file; if a cell ever needs an
 * embedded comma, swap this for a proper CSV parser like papaparse).
 */
function toRows(csvText: string): string[][] {
  const lines = csvText.split(/\r\n|\n/);
  return lines.map((line) => line.split(','));
}

function parseBlocks(rows: string[][]): ParsedBlock[] {
  const blocks: ParsedBlock[] = [];
  const n = rows.length;
  let i = 0;

  while (i < n) {
    if (cellAt(rows, i, 0) === 'Range') {
      const outriggersRaw = cellAt(rows, i + 1, 1);
      const slewRaw = cellAt(rows, i + 2, 1);
      const counterweightTonnes = parseFloat(cellAt(rows, i + 3, 1));
      const flyJibRaw = cellAt(rows, i + 4, 1);
      const hasJib = flyJibRaw !== 'No' && flyJibRaw !== '';
      const jibLengthM = hasJib ? parseFloat(flyJibRaw) : null;

      const headerStart = i + 6;
      const boomHeader = rows[headerStart] ?? [];
      const angleHeader = hasJib ? rows[headerStart + 1] ?? [] : null;
      const dataStart = hasJib ? headerStart + 2 : headerStart + 1;

      const columns: ParsedColumn[] = [];
      for (let c = 1; c < boomHeader.length; c++) {
        const parsedBoom = parseBoomHeaderCell(boomHeader[c] ?? '');
        if (!parsedBoom) continue;
        let offsetDeg: number | null = null;
        if (angleHeader) {
          offsetDeg = parseAngleHeaderCell(angleHeader[c] ?? '');
          if (offsetDeg === null) continue; // jib block but this column has no valid angle — skip
        }
        columns.push({ col: c, boomLengthM: parsedBoom.boomLengthM, orientation: parsedBoom.orientation, offsetDeg });
      }

      let r = dataStart;
      const dataRows: ParsedBlock['dataRows'] = [];
      while (r < n && !isBlankRow(rows, r) && cellAt(rows, r, 0) !== 'Range') {
        const radiusRaw = cellAt(rows, r, 0);
        if (radiusRaw !== '') {
          const radiusM = parseFloat(radiusRaw);
          if (!Number.isNaN(radiusM)) {
            const values = new Map<number, number>();
            for (const colInfo of columns) {
              const raw = cellAt(rows, r, colInfo.col);
              if (raw !== '') {
                const v = parseFloat(raw);
                if (!Number.isNaN(v)) values.set(colInfo.col, v);
              }
            }
            dataRows.push({ radiusM, values });
          }
        }
        r++;
      }

      blocks.push({
        outriggersFullyDeployed: outriggersRaw === 'Yes',
        slewRadiusDeg: slewRaw ? parseFloat(slewRaw) : null,
        counterweightTonnes,
        jibLengthM,
        columns,
        dataRows,
      });
      i = r + 1;
    } else {
      i++;
    }
  }
  return blocks;
}

/** Static metadata that isn't published in the duty-chart CSV itself
 *  (rope length, hook blocks, rigged weight) — supply per crane model. */
export interface CraneStaticMeta {
  craneModel: string;
  riggedWeightTonnes: number;
  baseCounterweightTonnes: number;
  maxLinePullTonnes: number;
  totalRopeLengthM: number;
  hookBlocks: HookBlockSpec[];
}

function buildCraneModel(blocks: ParsedBlock[], meta: CraneStaticMeta): CraneModel {
  const plainBlocks = blocks.filter((b) => b.jibLengthM === null);
  const jibBlocks = blocks.filter((b) => b.jibLengthM !== null);

  const counterweights: CounterweightConfig[] = plainBlocks.map((b) => {
    const boomConfigs: BoomConfig[] = b.columns.map((colInfo) => {
      const capacities: RadiusCapacityPoint[] = [];
      for (const row of b.dataRows) {
        const v = row.values.get(colInfo.col);
        if (v !== undefined) capacities.push({ radiusM: row.radiusM, capacityTonnes: v });
      }
      const entry: BoomConfig = { boomLengthM: colInfo.boomLengthM, capacities };
      if (colInfo.orientation === 'rear') entry.orientation = 'rear';
      return entry;
    });

    // attach jib configs for this counterweight
    const matchingJibBlocks = jibBlocks.filter((jb) => jb.counterweightTonnes === b.counterweightTonnes);
    for (const jb of matchingJibBlocks) {
      const boomsInJib = new Map<number, ParsedColumn[]>();
      for (const colInfo of jb.columns) {
        const list = boomsInJib.get(colInfo.boomLengthM) ?? [];
        list.push(colInfo);
        boomsInJib.set(colInfo.boomLengthM, list);
      }

      for (const [boomLengthM, colList] of boomsInJib.entries()) {
        const jibEntries: JibConfig[] = colList
          .slice()
          .sort((a, b2) => (a.offsetDeg ?? 0) - (b2.offsetDeg ?? 0))
          .map((colInfo) => {
            const capacities: RadiusCapacityPoint[] = [];
            for (const row of jb.dataRows) {
              const v = row.values.get(colInfo.col);
              if (v !== undefined) capacities.push({ radiusM: row.radiusM, capacityTonnes: v });
            }
            return {
              jibLengthM: jb.jibLengthM as number,
              offsetDeg: (colInfo.offsetDeg ?? 0) as 0 | 20 | 40,
              capacities,
            };
          });

        let target = boomConfigs.find((bc) => bc.boomLengthM === boomLengthM && bc.orientation === undefined);
        if (!target) {
          // jib chart references a boom length absent from the plain
          // chart — keep the data rather than dropping it
          target = { boomLengthM, capacities: [] };
          boomConfigs.push(target);
        }
        target.jibs = [...(target.jibs ?? []), ...jibEntries];
      }
    }

    const cwLabel = Number.isInteger(b.counterweightTonnes)
      ? String(b.counterweightTonnes)
      : String(b.counterweightTonnes).replace('.', '_');
    return {
      id: `cw_${cwLabel}t`,
      weightTonnes: b.counterweightTonnes,
      outriggersFullyDeployed: b.outriggersFullyDeployed,
      slewRadiusDeg: b.slewRadiusDeg ?? undefined,
      boomConfigs,
    };
  });

  return {
    craneModel: meta.craneModel,
    riggedWeightTonnes: meta.riggedWeightTonnes,
    baseCounterweightTonnes: meta.baseCounterweightTonnes,
    maxLinePullTonnes: meta.maxLinePullTonnes,
    totalRopeLengthM: meta.totalRopeLengthM,
    hookBlocks: meta.hookBlocks,
    counterweights,
  };
}

/** Main entry point: parse a duty-chart CSV's raw text into a
 *  fully-shaped CraneModel, given the crane's static (non-chart)
 *  metadata. */
export function parseCraneDutyChartCsv(csvText: string, meta: CraneStaticMeta): CraneModel {
  const rows = toRows(csvText);
  const blocks = parseBlocks(rows);
  return buildCraneModel(blocks, meta);
}
